### PR TITLE
auto save notes #29

### LINKED ISSUE #29: Auto save notes for every "x" seconds
Refer to the documentation of [notes schema](https://github.com/SankethBK/diaryvault/blob/master/docs.md#3-notes)

Notes bloc is defined in `lib/features/notes/presentation/bloc/notes`. We need to emit the `SaveNote` event with a timer of "x" seconds. 

Use `Timer.periodic` to emit this event in `lib/features/notes/presentation/pages/note_create_page.dart`. 

Next step would be to make "x" configurable by taking an input from settings

### PR DESCRIPTION
**Pull request by droidbg**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [X] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

#### Auto Save After 10 seconds.
This Feature will auto save note after X seconds, and only at the time of exit or save it will fetch notes. 

We are emiting the AutoSaveNote event with a timer of "x" seconds (currently it is specified as 10 seconds in code).
If Note is new, and auto save note is triggered multiple time. 
- First time AutoSaveNote event triggered it will save the note. 
- Next time it is triggered, it will update the note. 

Used Timer.periodic to emit AutoSaveNote event.

#### Enable Disable Toggle Feature for Auto Save added in Settings Page.
 Now user can go to setting page and enable or disable the auto save functionality. 

## Featured Covered in this PR
- [X] Auto saving note after every x seconds. 
- [X] Enable Disable Auto Save Toggle from settings.

## Related Tickets & Documents

- Closes #29 

## Tested Feature?

- [X] In Real Device. 
- [X] In Emulator

## Screenshot

Screenshot 2023-10-02 at 1 37 04 PM

### CODE REVIEW

File: lib/features/notes/presentation/bloc/notes/notes_state.dart [hunk 1]
```diff
@@ -197,3 +236,7 @@ class NoteDeletionSuccesful extends NotesState {
 class NoteDeletionFailed extends NotesState {
   const NoteDeletionFailed({required String id}) : super(id: id, safe: false);
 }
+
+class FetchAfterAutoSave extends NotesState {
+  const FetchAfterAutoSave({required String id}) : super(id: id, safe: false);
+}
```
[the remark this answers is not in the document]
Comment (droidbg) [reply lib/features/notes/presentation/bloc/notes/notes_state.dart:242]: If we auto save, and after that close notes using the cross icon, the already saved notes ( that are not fetched with every auto save) will not be displayed in home page, so we need another state to emit to fetch the notes.

File: lib/features/notes/presentation/pages/note_create_page.dart [hunk 2]
```diff
@@ -108,23 +127,23 @@ class _NoteCreatePageState extends State<NoteCreatePage> {
...
+                showToast("failed to fetch note");
               } else if (state is NotesSavingFailed) {
                 showToast("Failed to save note");
               } else if (state is NoteSavedSuccesfully) {
                 showToast(state.newNote!
                     ? "Note saved successfully"
                     : "Note updated successfully");
                 _routeToHome();
+              } else if (state is NoteAutoSavedSuccesfully) {
+                showToast("Note Auto Saved");
```
[the remark this answers is not in the document]
Comment (droidbg) [reply lib/features/notes/presentation/pages/note_create_page.dart:146]: okay will remove this toast.

File: lib/features/notes/presentation/pages/note_create_page.dart [hunk 3]
```diff
@@ -33,18 +36,34 @@ class _NoteCreatePageState extends State<NoteCreatePage> {
   late final NotesBloc notesBloc;
   late Image neonImage;
   late double topPadding = 0;
+  Timer? _saveTimer;
 
   @override
   void initState() {
     super.initState();
+    _initSaveTimer();
+  }
```
[the remark this answers is not in the document]
Comment (droidbg) [reply lib/features/notes/presentation/pages/note_create_page.dart]: @SankethBK did the fix for initState.

File: lib/features/notes/presentation/pages/note_create_page.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (droidbg) [reply lib/features/notes/presentation/pages/note_create_page.dart]: @SankethBK  removed the toast message, please check.

File: lib/features/notes/presentation/bloc/notes/notes_state.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (SankethBK) [reply lib/features/notes/presentation/bloc/notes/notes_state.dart:242]: Got it

**Pull request merged**